### ISSUE TITLE
feature request: indices(x) and either ix(x) or index(x)

### ISSUE THREAD

**Issue by pkoppstein** [stedolan/jq#380]
The definition of index(i) in jq is currently:

```
def index(i): .[i][0];  # Version 1.3
```

The utility of this escapes me.  Usually when one wants the index
of something in an array, one wants the first index of the item in the
array.  In any case, such a function can readily be implemented as
follows, though here it is named `ix` to avoid confusion:

```
# ix(x) returns the index of the first occurrence of x in the input array, or null
def ix(x): . as $in
  | reduce range(0;length) as $ix (null; if . == null and $in[$ix] == x then $ix else . end);
```

More generally, here is a function that generates a stream of all indices.  For speed, it inlines the call to select(_):

```
# All indices of x as a stream of integers
def indices(x): . as $in | range(0; length)
  | if ($in[.] == x) then . else empty end;
```

Obviously ix(x) could be defined in terms of indices(_) but at the cost of space and time efficiency.

These definitions could obviously be extended to strings, but it
seems to me that the added complexity would entail some cost in terms
of execution speed, but with very little added utility, since when working
with strings, one typically wants to find the index of one substring
(or the first occurrence of a regex match) within another.

So, in summary:
- if the existing (mysterious) "index" function cannot be renamed or
  ditched altogether, then I would propose adding ix(_) and indices(_) as defined above;
- otherwise, "ix" as defined above could be added under the name "index", along with indices(_).
